galsim/catalog.py

```diff
@@ -312,3 +317,334 @@ def itervalues(self):
+        sed_intermed = galsim.SED(os.path.join(galsim.meta_data.share_dir,'CWW_Sbc_ext.sed'))
+
+    # Get fit parameters.
+    params = record.field('bulgefit')
+    sparams = record.field('sersicfit')
+    # Get the status flag for the fits.
+    bstat = record.field('fit_status')[0]
+    sstat = record.field('fit_status')[4]
+    # Get the precomputed bulge-to-total ratio for the 2-component fits.
+    dvc_btt = record.field('fit_dvc_btt')
+    # Get the precomputed median absolute deviation for the 1- and 2-component fits.
+    bmad = record.field('fit_mad_b')
+    smad = record.field('fit_mad_s')
+
+    # First decide if we can / should use bulgefit, otherwise sersicfit.  This decision process
+    # depends on: the status flags for the fits, the bulge-to-total ratios (if near 0 or 1, just use
+    # single component fits), the sizes for the bulge and disk (if <=0 then use single component
+    # fits), the axis ratios for the bulge and disk (if <0.051 then use single component fits), and
+    # a comparison of the median absolute deviations to see which is better.
+    use_bulgefit = 1
+    if bstat<1 or bstat>4 or dvc_btt<0.1 or dvc_btt>0.9 or np.isnan(dvc_btt) or params[9]<=0 or \
+            params[1]<=0 or params[11]<0.051 or params[3]<0.051 or smad<bmad:
+        use_bulgefit = 0
+        # Then check if sersicfit is viable; if not, this object is a total failure:
+        if sstat<1 or sstat>4 or sparams[1]<=0 or sparams[0]<=0:
+            raise RuntimeError("Cannot make parametric model for this object")
+
+    # Now, if we're supposed to use the 2-component fits, get all the parameters.
+    if use_bulgefit:
+        # Bulge parameters.
+        bulge_q = params[11]
+        bulge_beta = params[15]*galsim.radians
+        bulge_hlr = 0.03*np.sqrt(bulge_q)*params[9]
+        bulge_flux = 2.0*np.pi*3.607*(bulge_hlr**2)*params[8]/0.03**2
+        # Disk parameters.
+        disk_q = params[3]
+        disk_beta = params[7]*galsim.radians
+        disk_hlr = 0.03*np.sqrt(disk_q)*params[1]
+        disk_flux = 2.0*np.pi*1.901*(disk_hlr**2)*params[0]/0.03**2
+        bfrac = bulge_flux/(bulge_flux+disk_flux)
+        # Make sure the bulge-to-total flux ratio is not nonsense.
+        if bfrac < 0 or bfrac > 1 or np.isnan(bfrac):
+            raise RuntimeError("Cannot make parametric model for this object")
+
+        # Then make the object.
+        if chromatic:
+            bulge = galsim.DeVaucouleurs(flux=1., half_light_radius=record['size_factor']*bulge_hlr) \
+                * sed_bulge.atRedshift(record['zphot']).withMagnitude(
+                record['mag_auto']-2.5*math.log10(bfrac*record['flux_factor']), _COSMOS_bandpass)
+            disk = galsim.Exponential(flux=1., half_light_radius=record['size_factor']*disk_hlr) \
+                * sed_disk.atRedshift(record['zphot']).withMagnitude(
+                record['mag_auto']-2.5*math.log10((1.-bfrac)*record['flux_factor']), _COSMOS_bandpass)
```

Comment:
Recommend using some temporary variables to make these easier to parse.  `bulge_mag`, `disk_mag`, etc.